### PR TITLE
[Performance]: Delayed Python pass module load

### PR DESCRIPTION
**Pull request by shaahji**
### [Performance]: Delayed Python module load
Introducing olive_config.json to encapsulates all the different global environment properties (like pass module configuration) to setup olive. Also, merged extra_dependencies.json into olive_config.json.

User can provide a alternative configuration file with the command line argument '--package-config'.

Instead of loading all modules (specifically passes) at launch, delay the load until after the run config is parsed and load only
the ones that are explicitly required.

Moved a few hard-coded dependency management into the olive config.

### Release Note:
Added support for extending available Olive passes and configuring each via data file (olive_config.json). The configuration file is provided to Olive run with command line param `--package-config`.

## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [x] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

File: olive/passes/onnx/vitis_ai/__init__.py
Comment (jambayk) [inline olive/passes/onnx/vitis_ai/__init__.py:5]: these are not passes and the vitis quantizer depends on this init file. I think it's okay to keep them here since the pass only imports then during run https://github.com/microsoft/Olive/blob/main/olive/passes/onnx/vitis_ai_quantization.py#L266

File: olive/workflows/run/__main__.py
Comment (jambayk) [inline olive/workflows/run/__main__.py:27]: is there a reason for preferring `-` over `_`?

File: olive/workflows/run/config.py
Comment (jambayk) [inline olive/workflows/run/config.py:27]: Can you use a different name like `PassModuleConfig` or something since we already have similar named configs like `RunPassConfig`, `FullPassConfig`, `PassConfigParam`?

File: olive/olive_config.json
Comment (guotuofeng) [inline olive/olive_config.json:4]: the stringlized module full path is hard to maintain. User might forget to update the module path when it is changed.

How could we make sure the path here is updated whenever module path changes?

File: docs/source/getstarted/installation.md
Comment (guotuofeng) [inline docs/source/getstarted/installation.md:66]: need refer to your private branch. after the PR is merged, update to main.

**Comment by guotuofeng** [comment]
> Instead of loading all modules (specifically passes) at launch, delay the load of modules until after the run config is parsed and load only the ones that are explicitly required.

What's the perf improvement compared with all passes are imported? From the time being, only the pass itself is imported and the dependencies only be imported until run. So, could I suppose the perf improvement is limit? What's the consideration of this PR?

**Review by guotuofeng — DISMISSED** [review]
need to understand the consideration or background about why we need the delay load PR.

File: olive/workflows/run/__main__.py
Comment (devang-ml) [inline olive/workflows/run/__main__.py:26]: We don't want to deprecate --config.
Is there a reason to introduce new --run-config ?

File: olive/workflows/run/__main__.py
Comment (devang-ml) [inline olive/workflows/run/__main__.py:15]: Description and name should be clearer that to indicate that this optional .json file, meant for advanced users only, lists location of passes you want to use in your --config.

File: olive/olive_config.json
Comment (devang-ml) [reply olive/olive_config.json:4]: As I understand, the change in path would require a PR and the pass would stop working if the string is not updated.

File: olive/workflows/run/__main__.py
Comment (shaahji) [reply olive/workflows/run/__main__.py:27]: I read about the convention to use '-' in the parameter which python converts it to '_' - justification being that it distinguishes the use cases and avoids accidental changes with operations like find/replace.

Unfortunately, I can't seem to find the actual source link again. :(

For the moment, both conventions are supported.

File: olive/workflows/run/config.py
Comment (shaahji) [reply olive/workflows/run/config.py:27]: In my opinion, this is the most appropriate name, especially since the other is named with prefix 'Run'. I would argue that the other classes should be named with the same prefix because they relate to the 'run'.

Regardless, how about `PassPackageConfig` since files are considered package in Python (borrowing the convention from `importlib.import_module` API where the second parameter is named `package`.)

File: olive/olive_config.json
Comment (shaahji) [reply olive/olive_config.json:4]: > As I understand, the change in path would require a PR and the pass would stop working if the string is not updated.

That's correct.

File: olive/workflows/run/__main__.py
Comment (shaahji) [reply olive/workflows/run/__main__.py:26]: Merely to explicitly differentiate between `olive-config` and `the other config`. In my opinion, think naming the other config explicitly i.e. prefixing with 'run-' adds value.

File: docs/source/getstarted/installation.md
Comment (shaahji) [reply docs/source/getstarted/installation.md:66]: I don't follow your comment. Please explain!

File: olive/workflows/run/__main__.py
Comment (shaahji) [reply olive/workflows/run/__main__.py:15]: Are you referring to using something else in place of 'olive-config'? I started off with calling it passes-config but then realized there are other scattered input files that can be merged into one and that can serve as a more *entire module config* rather than having to explain to user what each config does.

**Comment by shaahji** [comment]
> What's the perf improvement number compared with all passes are imported? From the time being, only the pass itself is imported and the dependencies only be imported until run. So, could I suppose the perf improvement is limited? What's the consideration of this PR?

> need to understand the consideration or background about why we need the delay load PR.

I don't think one would notice a significant performance gain with this change. All that loading/initialization is one time thing and not a hot path to really show up in any profiles. 

Primary motivation for this change -
* This approach gives the user a way to add his/her own pass and drive all with data.
* Load/initialization time might be small today with the limited passes we have but over time as the list of passes grow this might become significant. This change attempt to preemptively circumvent that scenario. 
* Passes can now freely include the necessary imports at package level rather than having to be localized in functions and such. Since the packages aren't loaded until a pass needs them, those dependencies won't matter.

@devang-ml can add more.

File: olive/passes/onnx/vitis_ai/__init__.py
Comment (shaahji) [reply olive/passes/onnx/vitis_ai/__init__.py:5]: Fixex.

File: docs/source/getstarted/installation.md
Comment (guotuofeng) [reply docs/source/getstarted/installation.md:66]: before the PR is merged, the main/olive/olive_config.json will not exist in main branch. As the result, the linkcheck would fail.

Comment (trajepl) [reply docs/source/getstarted/installation.md:66]: There is link checker in our CI pipeline and it will fail if the link is unreachable.https://github.com/microsoft/Olive/pull/358

File: olive/olive_config.json
Comment (trajepl) [inline olive/olive_config.json:4]: Not related to this PR:
This central passes config looks nice. If we think further, it might be a good place to put pass's attributes like:
1. which ep this pass can be run
2. which precision(float32/float16/int4) the passes can conduct.
3. which ort version it is supported
That might be helpful for auto optimization as well.
Like this kind of configs
https://github.com/microsoft/Olive/blob/main/olive/auto_optimizer/config_template/pass_capability.yaml

File: olive/olive_config.json
Comment (xiaoyu-work) [reply olive/olive_config.json:4]: We can raise a specific error if this path couldn't find the pass.

File: olive/workflows/run/__main__.py
Comment (xiaoyu-work) [reply olive/workflows/run/__main__.py:35]: Most people will not use advanced `--olive-config` but still `--config`. I think keeping `--config` will make Olive as simple to use as possible.

**Comment by xiaoyu-work** [comment]
Can you complete the 5th checklist:  add a description of this change to be included in the release notes.

File: docs/source/getstarted/installation.md
Comment (shaahji) [reply docs/source/getstarted/installation.md:66]: So, what's the workaround? Split the PR into two submissions? Or merge the PR while overriding/ignoring the lint checker?

File: olive/workflows/run/__main__.py
Comment (shaahji) [reply olive/workflows/run/__main__.py:24]: Added descriptive help text.

File: olive/workflows/run/__main__.py
Comment (shaahji) [reply olive/workflows/run/__main__.py:32]: To avoid having to import logger so soon in the launch process. 

Regardless, those messages are removed.

File: olive/olive_config.json
Comment (shaahji) [reply olive/olive_config.json:4]: >  If we think further, it might be a good place to put pass's attributes

Agreed. There is lot more that can be data driven allowing easier user configuration changes.

File: olive/systems/azureml/aml_pass_runner.py
Comment (jambayk) [inline olive/systems/azureml/aml_pass_runner.py:1]: Please also update https://github.com/microsoft/Olive/blob/main/olive/systems/python_environment/pass_runner.py

Comment (shaahji) [reply olive/systems/azureml/aml_pass_runner.py:1]: Done!

File: olive/olive_config.json
Comment (trajepl) [reply olive/olive_config.json:4]: > Don't we have passes that can be loaded only in specific environment like Azure or OpenVino?

Could you help explain more?

The UT function here I mean is like to verify all of `module_path` in  `olive_config.json` are right to the olive build-in passes. 
In my rough thoughts, we can just read the `olive_config.json` and start a looper to check if the `module_path` is right. 
I think it is quite simple, mostly? 

The UT I mean is just  for our build-in passes. Otherwise, if any typos of the `module_path` are checked into main, we have to start a new PR then wait CI and approval.

File: docs/source/getstarted/installation.md
Comment (trajepl) [reply docs/source/getstarted/installation.md:66]: Previously, we split the PR into two submissions.

File: olive/olive_config.json
Comment (shaahji) [reply olive/olive_config.json:4]: Done!

File: docs/source/getstarted/installation.md
Comment (shaahji) [reply docs/source/getstarted/installation.md]: Done! Removed docs changes from this PR and will follow up with another ones this is merged.

File: olive/systems/azureml/aml_pass_runner.py
Comment (guotuofeng) [inline olive/systems/azureml/aml_pass_runner.py:24]: did this import introduce circular import?

Comment (guotuofeng) [reply olive/systems/azureml/aml_pass_runner.py]: need we move OlivePackageConfig into olive_pass.py or pass_config.py so that the module dependencies between workflow and system could be decoupled?

File: olive/workflows/run/__main__.py
Comment (guotuofeng) [reply olive/workflows/run/__main__.py]: it seems the `--config` is renamed to `--run-config`? if this is the case, we need update all readme.md for all examples.

File: olive/workflows/run/__main__.py
Comment (guotuofeng) [reply olive/workflows/run/__main__.py:22]: If we change the `--config`, all the examples' readme need to be updated. Otherwise, better to keep using the original parameter name.

File: olive/systems/azureml/aml_pass_runner.py
Comment (shaahji) [reply olive/systems/azureml/aml_pass_runner.py]: I don't think this is an issue. All tests have passed successfully.

File: olive/workflows/run/__main__.py
Comment (shaahji) [reply olive/workflows/run/__main__.py:22]: `--config` was marked deprecated not removed. So using `--config` was still acceptable but `--run-config` was preferred. Regardless, that change has been completed reverted. `--config` stays and the new config parameter is called `--package-config`.

File: olive/workflows/run/__main__.py
Comment (shaahji) [reply olive/workflows/run/__main__.py]: `--run-config` is acceptable parameter in place of `--config`. New argument it explicitly named `--package-config`.

File: olive/systems/azureml/aml_pass_runner.py
Comment (guotuofeng) [reply olive/systems/azureml/aml_pass_runner.py]: I understand there is no issue from the time being. Not sure some body would change the SystemConfig depends on OlivePackageConfig in future. If not, there will no issues. Otherwise, we need avoid the circular imports.

Comment (shaahji) [reply olive/systems/azureml/aml_pass_runner.py]: Both _olive_pass.py_ and _pass_config.py_ seems like wrong place for `OlivePackageConfig`. How about creating a new config file specifically for this, say _olive/common/olive_config.py_. I could move the `PassModuleConfig` into _pass_config.py_.

Comment (guotuofeng) [reply olive/systems/azureml/aml_pass_runner.py]: sure, I am ok with your suggestion.

Comment (shaahji) [reply olive/systems/azureml/aml_pass_runner.py]: Done!

File: olive/common/package_config.py
Comment (shaahji) [reply olive/common/package_config.py:1]: Done!

**Comment by shaahji** [comment]
@guotuofeng Can you review the change please! Thx

File: setup.py
Comment (guotuofeng) [inline setup.py:34]: it seems this file could be removed?

Comment (shaahji) [reply setup.py:34]: Yes, as part of a follow up PR to avoid docs build from failing. WIP: #1051

File: olive/package_config.py
Comment (guotuofeng) [inline olive/package_config.py:2]: the file location seems be different with the PR https://github.com/microsoft/Olive/pull/1051. which PR is the final one?

Comment (shaahji) [reply olive/package_config.py:2]: #1051 is meant to be a follow up after this PR merges. I still have to rebase that PR.

**Comment by guotuofeng** [comment]
the PR looks good to me except above comments. I will let @devang-ml for more comments.

**Review by devang-ml — APPROVED** [review]

**Pull request merged**
